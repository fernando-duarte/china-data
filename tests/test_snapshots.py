"""
Snapshot testing for data pipelines using syrupy.

This module demonstrates snapshot testing to ensure data processing
output remains consistent across changes.
"""

from typing import Any, cast

import pandas as pd
import pytest

try:
    from syrupy import SnapshotAssertion
except ImportError:
    # Fallback for when syrupy is not available
    SnapshotAssertion = Any  # type: ignore[misc,assignment]


# Mock data processing functions for testing
def process_china_data_sample() -> dict[str, Any]:
    """Process a sample of China data for testing."""
    return {
        "gdp_growth": [6.1, 2.3, 8.1],
        "inflation": [2.9, 2.5, 0.9],
        "unemployment": [5.2, 5.9, 5.5],
        "years": [2019, 2020, 2021],
        "metadata": {
            "source": "test_data",
            "processed_at": "2025-01-01T00:00:00Z",
            "version": "1.0.0",
        },
    }


def process_economic_indicators() -> pd.DataFrame:
    """Process economic indicators for testing."""
    return pd.DataFrame(
        {
            "year": [2020, 2021, 2022],
            "gdp_usd": [14.34e12, 17.73e12, 17.95e12],
            "gdp_growth_rate": [2.3, 8.1, 3.0],
            "cpi_inflation": [2.5, 0.9, 2.0],
            "unemployment_rate": [5.9, 5.5, 5.4],
        }
    )


class TestDataProcessingSnapshots:
    """Test data processing output consistency using snapshots."""

    def test_china_data_processing_output(self, snapshot: SnapshotAssertion):
        """Ensure China data processing output remains consistent."""
        result = process_china_data_sample()
        assert result == snapshot

    def test_economic_indicators_dataframe(self, snapshot: SnapshotAssertion):
        """Ensure economic indicators DataFrame structure is consistent."""
        df = process_economic_indicators()

        # Convert DataFrame to dict for snapshot comparison
        result = {
            "columns": df.columns.tolist(),
            "dtypes": df.dtypes.astype(str).to_dict(),
            "shape": df.shape,
            "data": df.to_dict("records"),
        }

        assert result == snapshot

    def test_data_transformation_pipeline(self, snapshot: SnapshotAssertion):
        """Test complete data transformation pipeline output."""
        # Simulate a data transformation pipeline
        raw_data = {
            "gdp_nominal": [14.34e12, 17.73e12, 17.95e12],
            "population": [1.411e9, 1.412e9, 1.413e9],
            "years": [2020, 2021, 2022],
        }

        # Transform data (calculate GDP per capita)
        gdp_nominal = cast("list[float]", raw_data["gdp_nominal"])
        population = cast("list[float]", raw_data["population"])
        transformed = {
            "year": raw_data["years"],
            "gdp_per_capita": [
                gdp / pop for gdp, pop in zip(gdp_nominal, population, strict=False)
            ],
            "gdp_nominal_trillions": [gdp / 1e12 for gdp in gdp_nominal],
            "transformation_metadata": {
                "method": "gdp_per_capita_calculation",
                "currency": "USD",
                "base_year": 2020,
            },
        }

        assert transformed == snapshot

    @pytest.mark.parametrize(("year", "expected_records"), [(2020, 1), (2021, 2), (2022, 3)])
    def test_yearly_data_filtering(
        self, snapshot: SnapshotAssertion, year: int, expected_records: int
    ):
        """Test yearly data filtering with snapshots."""
        df = process_economic_indicators()

        # Filter data up to the specified year
        filtered_df = df[df["year"] <= year]

        result = {
            "filter_year": year,
            "record_count": len(filtered_df),
            "data": filtered_df.to_dict("records"),
        }

        assert result == snapshot
        assert len(filtered_df) == expected_records


class TestDataValidationSnapshots:
    """Test data validation output using snapshots."""

    def test_data_quality_report(self, snapshot: SnapshotAssertion):
        """Test data quality validation report structure."""
        df = process_economic_indicators()

        # Generate data quality report
        quality_report = {
            "total_records": len(df),
            "missing_values": df.isna().sum().to_dict(),
            "data_types": df.dtypes.astype(str).to_dict(),
            "numeric_summaries": {
                col: {
                    "min": float(df[col].min()),
                    "max": float(df[col].max()),
                    "mean": float(df[col].mean()),
                }
                for col in df.select_dtypes(include=["number"]).columns
            },
            "validation_timestamp": "2025-01-01T00:00:00Z",  # Fixed for testing
        }

        assert quality_report == snapshot

    def test_outlier_detection_results(self, snapshot: SnapshotAssertion):
        """Test outlier detection algorithm results."""
        df = process_economic_indicators()

        # Simple outlier detection using IQR method
        outliers = {}
        for col in df.select_dtypes(include=["number"]).columns:
            if col != "year":  # Skip year column
                q1 = df[col].quantile(0.25)
                q3 = df[col].quantile(0.75)
                iqr = q3 - q1
                lower_bound = q1 - 1.5 * iqr
                upper_bound = q3 + 1.5 * iqr

                outlier_mask = (df[col] < lower_bound) | (df[col] > upper_bound)
                outlier_indices = df[outlier_mask].index.tolist()
                outliers[col] = {
                    "outlier_count": int(outlier_mask.sum()),
                    "outlier_indices": outlier_indices,
                    "bounds": {"lower": float(lower_bound), "upper": float(upper_bound)},
                }

        result = {
            "outlier_analysis": outliers,
            "total_outliers": sum(info["outlier_count"] for info in outliers.values()),
        }

        assert result == snapshot


# Configuration for snapshot testing
@pytest.fixture
def snapshot_config():
    """Configure snapshot testing behavior."""
    return {
        "update_snapshots": False,  # Set to True to update snapshots
        "include_metadata": True,
    }
