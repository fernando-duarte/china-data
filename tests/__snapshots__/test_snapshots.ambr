# serializer version: 1
# name: TestDataProcessingSnapshots.test_china_data_processing_output
  dict({
    'gdp_growth': list([
      6.1,
      2.3,
      8.1,
    ]),
    'inflation': list([
      2.9,
      2.5,
      0.9,
    ]),
    'metadata': dict({
      'processed_at': '2025-01-01T00:00:00Z',
      'source': 'test_data',
      'version': '1.0.0',
    }),
    'unemployment': list([
      5.2,
      5.9,
      5.5,
    ]),
    'years': list([
      2019,
      2020,
      2021,
    ]),
  })
# ---
# name: TestDataProcessingSnapshots.test_data_transformation_pipeline
  dict({
    'gdp_nominal_trillions': list([
      14.34,
      17.73,
      17.95,
    ]),
    'gdp_per_capita': list([
      10163.004961020553,
      12556.657223796034,
      12703.4677990092,
    ]),
    'transformation_metadata': dict({
      'base_year': 2020,
      'currency': 'USD',
      'method': 'gdp_per_capita_calculation',
    }),
    'year': list([
      2020,
      2021,
      2022,
    ]),
  })
# ---
# name: TestDataProcessingSnapshots.test_economic_indicators_dataframe
  dict({
    'columns': list([
      'year',
      'gdp_usd',
      'gdp_growth_rate',
      'cpi_inflation',
      'unemployment_rate',
    ]),
    'data': list([
      dict({
        'cpi_inflation': 2.5,
        'gdp_growth_rate': 2.3,
        'gdp_usd': 14340000000000.0,
        'unemployment_rate': 5.9,
        'year': 2020,
      }),
      dict({
        'cpi_inflation': 0.9,
        'gdp_growth_rate': 8.1,
        'gdp_usd': 17730000000000.0,
        'unemployment_rate': 5.5,
        'year': 2021,
      }),
      dict({
        'cpi_inflation': 2.0,
        'gdp_growth_rate': 3.0,
        'gdp_usd': 17950000000000.0,
        'unemployment_rate': 5.4,
        'year': 2022,
      }),
    ]),
    'dtypes': dict({
      'cpi_inflation': 'float64',
      'gdp_growth_rate': 'float64',
      'gdp_usd': 'float64',
      'unemployment_rate': 'float64',
      'year': 'int64',
    }),
    'shape': tuple(
      3,
      5,
    ),
  })
# ---
# name: TestDataProcessingSnapshots.test_yearly_data_filtering[2020-1]
  dict({
    'data': list([
      dict({
        'cpi_inflation': 2.5,
        'gdp_growth_rate': 2.3,
        'gdp_usd': 14340000000000.0,
        'unemployment_rate': 5.9,
        'year': 2020,
      }),
    ]),
    'filter_year': 2020,
    'record_count': 1,
  })
# ---
# name: TestDataProcessingSnapshots.test_yearly_data_filtering[2021-2]
  dict({
    'data': list([
      dict({
        'cpi_inflation': 2.5,
        'gdp_growth_rate': 2.3,
        'gdp_usd': 14340000000000.0,
        'unemployment_rate': 5.9,
        'year': 2020,
      }),
      dict({
        'cpi_inflation': 0.9,
        'gdp_growth_rate': 8.1,
        'gdp_usd': 17730000000000.0,
        'unemployment_rate': 5.5,
        'year': 2021,
      }),
    ]),
    'filter_year': 2021,
    'record_count': 2,
  })
# ---
# name: TestDataProcessingSnapshots.test_yearly_data_filtering[2022-3]
  dict({
    'data': list([
      dict({
        'cpi_inflation': 2.5,
        'gdp_growth_rate': 2.3,
        'gdp_usd': 14340000000000.0,
        'unemployment_rate': 5.9,
        'year': 2020,
      }),
      dict({
        'cpi_inflation': 0.9,
        'gdp_growth_rate': 8.1,
        'gdp_usd': 17730000000000.0,
        'unemployment_rate': 5.5,
        'year': 2021,
      }),
      dict({
        'cpi_inflation': 2.0,
        'gdp_growth_rate': 3.0,
        'gdp_usd': 17950000000000.0,
        'unemployment_rate': 5.4,
        'year': 2022,
      }),
    ]),
    'filter_year': 2022,
    'record_count': 3,
  })
# ---
# name: TestDataValidationSnapshots.test_data_quality_report
  dict({
    'data_types': dict({
      'cpi_inflation': 'float64',
      'gdp_growth_rate': 'float64',
      'gdp_usd': 'float64',
      'unemployment_rate': 'float64',
      'year': 'int64',
    }),
    'missing_values': dict({
      'cpi_inflation': 0,
      'gdp_growth_rate': 0,
      'gdp_usd': 0,
      'unemployment_rate': 0,
      'year': 0,
    }),
    'numeric_summaries': dict({
      'cpi_inflation': dict({
        'max': 2.5,
        'mean': 1.8,
        'min': 0.9,
      }),
      'gdp_growth_rate': dict({
        'max': 8.1,
        'mean': 4.466666666666666,
        'min': 2.3,
      }),
      'gdp_usd': dict({
        'max': 17950000000000.0,
        'mean': 16673333333333.334,
        'min': 14340000000000.0,
      }),
      'unemployment_rate': dict({
        'max': 5.9,
        'mean': 5.6000000000000005,
        'min': 5.4,
      }),
      'year': dict({
        'max': 2022.0,
        'mean': 2021.0,
        'min': 2020.0,
      }),
    }),
    'total_records': 3,
    'validation_timestamp': '2025-01-01T00:00:00Z',
  })
# ---
# name: TestDataValidationSnapshots.test_outlier_detection_results
  dict({
    'outlier_analysis': dict({
      'cpi_inflation': dict({
        'bounds': dict({
          'lower': 0.24999999999999978,
          'upper': 3.45,
        }),
        'outlier_count': 0,
        'outlier_indices': list([
        ]),
      }),
      'gdp_growth_rate': dict({
        'bounds': dict({
          'lower': -1.6999999999999997,
          'upper': 9.899999999999999,
        }),
        'outlier_count': 0,
        'outlier_indices': list([
        ]),
      }),
      'gdp_usd': dict({
        'bounds': dict({
          'lower': 13327500000000.0,
          'upper': 20547500000000.0,
        }),
        'outlier_count': 0,
        'outlier_indices': list([
        ]),
      }),
      'unemployment_rate': dict({
        'bounds': dict({
          'lower': 5.075,
          'upper': 6.075,
        }),
        'outlier_count': 0,
        'outlier_indices': list([
        ]),
      }),
    }),
    'total_outliers': 0,
  })
# ---
